File: src/Butil/Bit.Butil/Publics/TextFragment.cs
using System.Diagnostics.CodeAnalysis;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps <see href="https://developer.mozilla.org/en-US/docs/Web/URI/Fragment/Text_fragments">text fragments</see>:
/// URLs that scroll to and highlight a <b>phrase</b> rather than an anchor - <c>#:~:text=…</c>.
/// </summary>
/// <remarks>
/// The point is deep-linking into content you don't control the markup of: no <c>id</c> has to
/// exist, and the link keeps working when the page is re-flowed, as long as the words survive.
/// This is what a browser's "Copy link to highlight" produces.
/// <br/>
/// <b>How the browser treats it:</b> everything after <c>:~:</c> is a <i>fragment directive</i>, and
/// the browser strips it off the URL while navigating - so it is gone from <c>location</c> before any
/// page script runs, and <see cref="Parse"/> reads a URL you still hold rather than the current one.
/// A directive is only acted on during a navigation, so <see cref="Navigate"/> is what makes the
/// browser scroll; writing the URL into history does nothing.
/// <br/>
/// Style the result with the <c>::target-text</c> pseudo-element. Chromium and Safari implement
/// this; Firefox does not, and there the link simply loads the page without scrolling.
/// </remarks>
[ButilService(typeof(TextFragment))]
public class TextFragment(IJSRuntime js)
{
    /// <summary>True when the runtime exposes <c>document.fragmentDirective</c>.</summary>
    /// <remarks>
    /// A false here doesn't make the links unusable - they just load the page without scrolling, so
    /// this is a "will the highlight happen" check rather than a gate on generating them.
    /// <br/>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.textFragment.isSupported");

    /// <summary>
    /// Encodes directives into the fragment string, <c>:~:text=…</c>, with the escaping the spec
    /// requires.
    /// </summary>
    /// <returns>The fragment, or an empty string when no directive had a <see cref="TextFragmentDirective.Start"/>.</returns>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(TextFragmentDirective))]
    public ValueTask<string> Build(params TextFragmentDirective[] directives)
        => js.Invoke<string>("BitButil.textFragment.build", (object)directives);

    /// <summary>
    /// Appends directives to a URL, replacing any it already carries and keeping an ordinary
    /// <c>#anchor</c> in front of them.
    /// </summary>
    /// <param name="url">The URL to link to.</param>
    /// <param name="directives">What to highlight. More than one highlights more than one place.</param>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(TextFragmentDirective))]
    public ValueTask<string> BuildUrl(string url, params TextFragmentDirective[] directives)
        => js.Invoke<string>("BitButil.textFragment.buildUrl", url, directives);

    /// <summary>
    /// Reads the directives out of a URL - the inverse of <see cref="BuildUrl"/>, for inspecting or
    /// rewriting a highlight link before navigating to it.
    /// </summary>
    /// <param name="url">A URL that may carry a <c>:~:text=…</c> fragment directive.</param>
    /// <returns>The directives, or an empty array when the URL carries none.</returns>
    /// <remarks>
    /// This takes a URL rather than reading the current one because the current one no longer has a
    /// directive to read: the browser strips the fragment directive off while navigating, so it is
    /// gone from <c>location.hash</c> <i>and</i> from <c>location.href</c> by the time any script
    /// runs, and <c>document.fragmentDirective</c> deliberately exposes nothing. To know what a
    /// visitor was linked to, pass the URL through your own app - a query parameter, say - or parse
    /// the link before you navigate to it.
    /// </remarks>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(TextFragmentDirective))]
    public ValueTask<TextFragmentDirective[]> Parse(string url)
        => js.Invoke<TextFragmentDirective[]>("BitButil.textFragment.parse", url);

    /// <summary>
    /// Turns the user's current selection into a directive - the "copy link to this passage" button.
    /// </summary>
    /// <returns>The directive, or null when nothing is selected.</returns>
    /// <remarks>
    /// A short selection becomes an exact match; a long one becomes a start/end range, so the URL
    /// stays a reasonable length without matching any less precisely.
    /// </remarks>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(TextFragmentDirective))]
    public ValueTask<TextFragmentDirective?> FromSelection()
        => js.Invoke<TextFragmentDirective?>("BitButil.textFragment.fromSelection");

    /// <summary>
    /// Navigates to a URL so the browser acts on its text directive.
    /// </summary>
    /// <param name="url">A URL built by <see cref="BuildUrl"/>.</param>
    /// <param name="replace">Replace the current history entry instead of pushing a new one.</param>
    /// <remarks>
    /// This is a real navigation - in a Blazor app that means a full reload unless the URL is the
    /// current page with a different fragment. Scroll-to-text has no scripted equivalent: the
    /// browser only performs it as part of navigating.
    /// </remarks>
    public ValueTask Navigate(string url, bool replace = false)
        => js.InvokeVoid("BitButil.textFragment.navigate", url, replace);
}

File: src/Butil/Bit.Butil/Scripts/textFragment.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    // A text directive lives after `:~:` in the fragment, which is the "fragment directive" the
    // browser strips out of location.hash before any page code sees it.
    const DELIMITER = ':~:';

    // encodeURIComponent leaves `-`, `,` and `&` alone, and all three are structural inside a text
    // directive: `-` separates prefix/suffix, `,` separates the parts, `&` separates directives.
    function encodePart(value: string) {
        return encodeURIComponent(value ?? '')
            .replace(/-/g, '%2D')
            .replace(/,/g, '%2C')
            .replace(/&/g, '%26');
    }

    function encode(directive: any) {
        if (!directive?.start) return '';

        let text = '';
        if (directive.prefix) text += `${encodePart(directive.prefix)}-,`;
        text += encodePart(directive.start);
        if (directive.end) text += `,${encodePart(directive.end)}`;
        if (directive.suffix) text += `,-${encodePart(directive.suffix)}`;

        return `text=${text}`;
    }

    function decode(raw: string) {
        // text=[prefix-,]start[,end][,-suffix]
        const body = raw.slice('text='.length);
        const parts = body.split(',');
        const result: any = { prefix: '', start: '', end: '', suffix: '' };

        let index = 0;
        if (parts[index]?.endsWith('-')) {
            result.prefix = decodeURIComponent(parts[index].slice(0, -1));
            index++;
        }
        if (parts[parts.length - 1]?.startsWith('-')) {
            result.suffix = decodeURIComponent(parts[parts.length - 1].slice(1));
            parts.length -= 1;
        }
        result.start = decodeURIComponent(parts[index] ?? '');
        if (parts[index + 1] !== undefined) result.end = decodeURIComponent(parts[index + 1]);

        return result;
    }

    function fragmentOf(url: string) {
        const hash = url.indexOf('#');
        return hash < 0 ? '' : url.slice(hash + 1);
    }

    butil.textFragment = {
        isSupported() { return 'fragmentDirective' in document; },
        build(directives: any[]) {
            const encoded = (directives ?? []).map(encode).filter(d => !!d);
            return encoded.length ? `${DELIMITER}${encoded.join('&')}` : '';
        },
        buildUrl(url: string, directives: any[]) {
            const fragment = butil.textFragment.build(directives);
            if (!fragment) return url;

            // Only the fragment can hold a directive, so the split is on the first `#`: a `:~:` in
            // the path or the query is ordinary text and has to survive untouched. Anything already
            // after `:~:` is a fragment directive of its own and is replaced; an ordinary `#anchor`
            // in front of it is kept.
            const hash = url.indexOf('#');
            const head = hash < 0 ? url : url.slice(0, hash);
            const anchor = (hash < 0 ? '' : url.slice(hash + 1)).split(DELIMITER)[0];

            return `${head}#${anchor}${fragment}`;
        },
        parse(url: string) {
            const fragment = fragmentOf(url ?? '');
            const start = fragment.indexOf(DELIMITER);
            if (start < 0) return [];

            return fragment.slice(start + DELIMITER.length)
                .split('&')
                .filter(part => part.startsWith('text='))
                .map(decode);
        },
        fromSelection() {
            const selection = window.getSelection();
            const text = selection?.toString().trim() ?? '';
            if (!text) return null;

            // A long selection makes an unwieldy URL and matches no better, so past a threshold the
            // directive becomes a start/end range - which is exactly what the spec's `,` form is for.
            const words = text.split(/\s+/);
            if (words.length <= 8) return { prefix: '', start: text, end: '', suffix: '' };

            return {
                prefix: '',
                start: words.slice(0, 4).join(' '),
                end: words.slice(-4).join(' '),
                suffix: ''
            };
        },
        navigate(url: string, replace: boolean) {
            // A text directive is only acted on during a navigation, and only a same-document one
            // if the fragment actually changes - hence assign/replace rather than a hash write.
            if (replace) window.location.replace(url);
            else window.location.assign(url);
        }
    };
}(BitButil));
